File: 1001-2000/1019. Next Greater Node In Linked List.java
// Medium
// LinkedList, Stack
// Time:O(n),Space:O(n)
// https://leetcode.cn/problems/next-greater-node-in-linked-list/

import java.util.*;
class Solution {
  // Convert LinkedList to List first
  // Use monotonic decreasing stack to find next greater elements
  //   - if stack'top is smaller, update res[stack.pop()] as cur element
  public int[] nextLargerNodes(ListNode head) {
      List<Integer> list = new ArrayList<>();
      ListNode cur = head;
      while (cur != null) {
          list.add(cur.val);
          cur = cur.next;
      }
      int[] res = new int[list.size()];
      Stack<Integer> stack = new Stack<>();

      for (int i = 0; i < list.size(); i++) {
          while (!stack.isEmpty() && list.get(stack.peek()) < list.get(i)) {
              res[stack.pop()] = list.get(i);
          }
          stack.push(i);
      }
      return res;
  }
}

File: 1001-2000/1019. Next Greater Node In Linked List.py
# Time:O(n), Space:O(n)

from typing import List, Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        # turn linkedlist to arr
        vals = []
        cur = head
        while cur:
            vals.append(cur.val)
            cur = cur.next
        
        n = len(vals)
        ans = [0] * n
        stack = []

        for i in range(n):
            # cur val > the val corresponding to stack top
            while stack and vals[stack[-1]] < vals[i]:
                idx = stack.pop()
                ans[idx] = vals[i]
            stack.append(i)
        
        return ans
    